purchases/src/main/kotlin/com/revenuecat/purchases/Package.kt

```diff
@@ -7,14 +7,41 @@ import com.revenuecat.purchases.models.StoreProduct
  * @property identifier Unique identifier for this package. Can be one a predefined package type or a custom one.
  * @property packageType Package type for the product. Will be one of [PackageType].
  * @property product [StoreProduct] assigned to this package.
- * @property offering offering this package was returned from.
+ * @property offering offering Id this package was returned from.
+ * @property presentedOfferingContext [PresentedOfferingContext] from which this package was obtained.
  */
 data class Package(
     val identifier: String,
     val packageType: PackageType,
     val product: StoreProduct,
-    val offering: String,
-)
+    val presentedOfferingContext: PresentedOfferingContext,
+) {
+    @Deprecated(
+        "Use constructor with presentedOfferingContext instead",
+        ReplaceWith(
+            "Package(identifier, packageType, product, offering, " +
+                "PresentedOfferingContext(offeringIdentifier = offering))",
+        ),
+    )
+    constructor(
+        identifier: String,
+        packageType: PackageType,
+        product: StoreProduct,
+        offering: String,
+    ) : this(
+        identifier,
+        packageType,
+        product,
+        PresentedOfferingContext(offeringIdentifier = offering),
+    )
+
+    @Deprecated(
+        "Use presentedOfferingContext.offeringIdentifier instead",
+        ReplaceWith("presentedOfferingContext.offeringIdentifier"),
+    )
+    val offering: String
+        get() = presentedOfferingContext.offeringIdentifier ?: ""
```

Comment:
I made the `presentedOfferingContext` have a nullable `offeringIdentifier` instead of making the `presentedOfferingContext` nullable in StoreProduct/SubscriptionOption in case we could have non-offering related context, which is probably not the case right now... But this forces us to default to something here as a fallback. It should never be null right now though in the `Package`.

Reply:
I like this! I think this makes sense 👍

Reply:
hmm I think it makes sense for `StoreProduct/SubscriptionOption`, but with this there's no way to indicate that a package will always have an offering and a package is always going to have an offering. Do we need to replace `offering` in `Package` with `presentedOfferingContext`? I think we can leave this as is `offering: String` and only replace `presentedOfferingIdentifier` in `StoreProduct` and `SubscriptionOption`? We access the `presentedOfferingContext` through the products, and not the packages.

Reply:
Hmm it's a good point... the only difference is going to be that the offering is not nullable in the Package... However, It feels nicer to me if we have the same format in all 3 places... If we don't have this data at the `Package` level we will need to access it through the StoreProduct when purchasing, which seems slightly less clean... 

What do you think about leaving the offering property undeprecated here in the Package? It would be duplicated with the one in `presentedOfferingContext` but that way we can keep the non-nullability here

Reply:
I guess my question is if we need a `presentedOfferingContext` at all in `Package`. Is it ever going to be different than the `offering` value? I guess we can add a different way of getting Packages in the future, which would indeed make offering not nullable and create the need of a `presentedOfferingContext`, but maybe we can add that property whenever we need. I just think adding it right now makes the API confusing and doesn't add value. Maybe I am missing context

Reply:
Actually... What if we make it a sealed class that only has an implementation right now? That way we can keep the original nullabilities. Can you remind me what other type of data we are adding to the `PresentedOfferingContext`?

Reply:
@joshdholtz I was chatting live with @vegaro about this one... In the end, we believe it would be better to make the `offeringIdentifier` not nullable inside the `PresentedOfferingContext` so we can make the `PresentedOfferingContext` not nullable in `Package` but nullable in `StoreProduct`/`SubscriptionOption`. Do you think we will (now or in the future) have data like this without having an offering id? 

My original reason to make it nullable was to provide more flexibility in the future in case we have some of this data without an offering id, but I don't think that would happen for now at least, so it makes sense to better reflect the current model.

Let us know what you think!

Reply:
Addressed in https://github.com/RevenueCat/purchases-android/pull/1612/commits/9eaf8a42fb199168f586f2f9c1e80b53f46860ca